### PR TITLE
Greg

### PR DESCRIPTION
**Pull request by gpric024**
Made changes and new test class based on previous feedback

### CODE REVIEW

File: DyslexiaInjector.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+        Injects dyslexia into the dataset by swapping words and letters.
+    get_homophones(word)
+        Returns the homophones of a word
+    get_confusing_letters(letter)
+        Returns the confusing letters of a letter
+    injector(sentence, p_homophone, p_letter)
+        Injects dyslexia into a sentence with a given probability
+    """
+    def __init__(self, load: DataLoader, homophone_path = "data/homophones_dict.pickle",
+                confusing_letters_path = "data/confusing_letters_dict.pickle", confusing_words_path="data/pedler_dict.pickle", seed = 42):
```
Comment (Shaomei) [inline DyslexiaInjector.py]: nit: line too long, try to keep line length within 80 characters for better readability.

Similar for others long lines below.

File: DataLoader.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    save_as_txt(path):
+        Saves data as a text file to specified path
+    save_as_csv(path):
+        Saves data as a csv file to specified path
+    get_data():
+        Returns data
+    create_deepcopy():
+        Returns a deep copy of the class instance
+    get_name():
+        returns name of the dataset (dataset_name)
```
Comment (Shaomei) [inline DataLoader.py:29]: Can you add a section for "Usage", and show a few examples of how to use this class?

File: DyslexiaInjector.py [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+        Loads the confusing letters from a pickle file
+    injection_swap(p_start=0, p_end=1, step_size=0.1, save_path="", save_format="both")
+        Injects dyslexia into the dataset by swapping words and letters.
+    get_homophones(word)
+        Returns the homophones of a word
+    get_confusing_letters(letter)
+        Returns the confusing letters of a letter
+    injector(sentence, p_homophone, p_letter)
+        Injects dyslexia into a sentence with a given probability
+    """
```
Comment (Shaomei) [inline DyslexiaInjector.py:44]: Same as above. This class is getting quite complicated, could be helpful to provide a few examples of how to use it here.

File: TestInjector.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    
+    def test_homophone_swapper(self):
+        original_word = "Capital,"
+        word = "capital"
+        apostrophe = False
+        out_word, apostrophe = self.injector.homophone_swapper(original_word, word)
+        #check that first letter is capitalized
+        self.assertEqual(out_word[0].isupper(), True)
+        #check that words are different
+        self.assertNotEqual(out_word, word)
```
Comment (Shaomei) [inline TestInjector.py:30]: What do we expect the out_word to be? Can we use another assert specify it here, too?

File: TestInjector.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        swapped_count = 0
+        for i in range(len(out_word)):
+            if original_word.strip('".,?!:;()').strip("'")[i].isupper():
+                self.assertEqual(out_word[i].isupper(), True)
+            else:
+                self.assertEqual(out_word[i].islower(), True)
+            if out_word[i].lower() != word[i].lower():
+                swapped_count += 1
+        #checking if correct number of letters were swapped
+        self.assertEqual(swapped_count, letters_swapped, f"actual_swapped_count: {swapped_count} | given_letters_swapped: {letters_swapped}")
```
Comment (Shaomei) [inline TestInjector.py:57]: Besides checking letters_swapped, can we also check whether the out_word is what we expected to be?

File: TestInjector.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    
+    def test_confusing_word_injector(self):
+        original_word = "Back"
+        word = "back"
+        out_word = self.injector.confusing_word_injector(original_word, word)
+        print(f"original word: {original_word} | out_word: {out_word}")
+        self.assertNotEqual(out_word, word)
```
Comment (Shaomei) [inline TestInjector.py:66]: So what is the expected out_word?

File: TestSentence.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
+
+import unittest
+#objective of this test is to make sure that the sentences in the dataset are clean
```
Comment (Shaomei) [inline TestSentence.py]: Hmm, checking the sentences in the dataset does not sound like a unit test. 

Maybe we should just make it into a PrecheckDataset class and tell people (including ourselves) to run it before the injection.

**Review by Shaomei — APPROVED** [review]
Looking good overall. 

You can merge after addressing my comments, no need to review again.

File: TestInjector.py (hunk 6 above)
Comment (gpric024) [reply TestInjector.py:66]: added regex to get the proper output

File: TestSentence.py (hunk 7 above)
Comment (gpric024) [reply TestSentence.py]: I do have a fix_formatting method in the dataloader class which handles all these cases. This is more to ensure that nothing has slipped through the cracks.  For now, I think I can remove it.

**Pull request merged**